Allow composition of pure-ASCII strings in the mode line

author	Eli Zaretskii <eliz@gnu.org>	2020-02-08 15:41:36 +0200

* src/composite.c (Fcomposition_get_gstring): Allow unibyte strings if they are pure ASCII, by copying text into a multibyte string.

--- emacs-27.2/src/composite.c.orig	2021-01-28 18:52:38.000000000 +0000
+++ emacs-27.2/src/composite.c	2021-12-09 14:11:21.026293233 +0000
@@ -1769,7 +1769,18 @@
       CHECK_STRING (string);
       validate_subarray (string, from, to, SCHARS (string), &frompos, &topos);
       if (! STRING_MULTIBYTE (string))
-	error ("Attempt to shape unibyte text");
+      {
+        ptrdiff_t i;
+
+        for (i = SBYTES (string) - 1; i >= 0; i--)
+          if (!ASCII_CHAR_P (SREF (string, i)))
+            error ("Attempt to shape unibyte text");
+        /* STRING is a pure-ASCII string, so we can convert it (or,
+           rather, its copy) to multibyte and use that thereafter.  */
+        Lisp_Object string_copy = Fconcat (1, &string);
+        STRING_SET_MULTIBYTE (string_copy);
+        string = string_copy;
+      }
       frombyte = string_char_to_byte (string, frompos);
     }
 
